carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * CPP File containing Mobility Operation Message method implementations
+ */
+
+#include "MobilityOperation_Message.h"
+
+namespace Message_cpp
+{
+    boost::optional<cav_msgs::MobilityOperation> Mobility_Operation::decode_mobility_operation_message(std::vector<uint8_t>& binary_array){
+        
+        cav_msgs::MobilityHeader header;
+        cav_msgs::MobilityOperation output;
+        //decode results - stored in binary_array
+        asn_dec_rval_t rval;
+        MessageFrame_t* message=0;
+        
+        //copy from vector to array         
+        auto len=binary_array.size();    
+        
+        uint8_t buf[len];             
+        for(auto i=0;i < len;i++){
+            buf[i]=binary_array[i];
+        }
+        //use asn1c lib to decode
+        
+        rval=uper_decode(0, &asn_DEF_MessageFrame,(void **) &message, buf, len, 0, 0);
+         
+        //if decode success
+        if(rval.code==RC_OK){
+           
+            //convert strategy from char array to string (TestMessage03 for MobilityOperation)
+            std::string sender_id, recipient_id, sender_bsm_id, plan_id,timestamp_string, strategy,strategy_params;
+            uint64_t timestamp;
+            //get sender id
+            auto str_len=message->value.choice.TestMessage03.header.hostStaticId.size;
+            if(str_len<=STATIC_ID_MAX_LENGTH && str_len!=0)
+            {
+                for(auto i=0;i<str_len;i++){
+                    sender_id +=message->value.choice.TestMessage03.header.hostStaticId.buf[i];
```

Comment:
Is this correct to how the original implementation was? I'm trying to wrap my head around the implementation here. A static ID like this is a 4-byte value (if I recall correctly), adding the 4 bytes together individually != the original 4 byte string: (e.g. 11001010<sub>2</sub> = 202<sub>10</sub>, 1100<sub>2</sub> + 1010<sub>2</sub> = 10110<sub>2</sub> or 22<sub>10</sub>).

To correctly do this we'd need to bitshift and add: `1100 << 8 + 1010 = 11001010`.

If this is how the Java code implemented it, we can think on that this may be a load-bearing defect.